
|                                                                                           |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                |
| ----------------------------------------------------------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| What is the syntax<br>when create a new <br>syntax                                        | ![[Pasted image 20240319131806.png]]<br>this is called throw exception<br><br>only instance of throwable are allowed to throw<br>                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              |
| When is required to <br>create a new <br>exception                                        | ![[Pasted image 20240319131947.png]]<br><br>This is solution for this exception<br><br>![[Pasted image 20240319132544.png]]<br><br>Explain this case <br>demo-22<br><br>A new class is created called " InvalidCustomerIdExcepition" extending with the exception.<br><br>continue....                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         |
| What is the <br>use of throw<br>exception                                                 | 1. Throw exception is used to create the new exception when required<br><br>2. Re throw                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        |
| Is  any compiler error<br>encountered<br>when the <br>created exception is <br>not thrown | ![[Pasted image 20240319133234.png]]<br>Nothing happened when the exception is not thrown<br><br>![[Pasted image 20240319133330.png]]<br>now this should be specify or catch                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   |
| What are the unreachable<br>statement when<br>use the throw                               | the below line after the throw statement are unreachable statements<br><br>![[Pasted image 20240319133817.png]]<br>                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            |
| What can be thrown                                                                        | ![[Pasted image 20240319134017.png]]<br><br>Only instance of throwable are allowed to throw.In here app is not a instance of throwalbe. So this is not allowed to throw. (throw and thorws are two difference things. This rule is not for the throws)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         |
| What is the rethrow<br><br>demo-24                                                        | ![[Pasted image 20240319134623.png]]<br><br>In here try-catch in myMethod2( ) failed to handle the error. But it throws the exceptio again. Now myMethod2( ) should catch or specify this exception                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            |
| What does mean the exception chain                                                        | Wrap up exception from a another exception type is called the exception chain<br><br>![[Pasted image 20240319135323.png]]<br>in myMethod1( ) the Exception e has been chain with the RuntimeException<br><br>In here the exception is not handled by the runtime exception<br><br>![[Pasted image 20240319135642.png]]<br>Now this is handled in the main method( ) handling the myMethod1 ( )                                                                                                                                                                                                                                                                                                                                                                                                                                                                 |
| How to print the <br>stack trace explicitly                                               | ![[Pasted image 20240319135804.png]]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           |
| What is the reason for only<br>error is occurred <br>in the method3()                     | ![[Pasted image 20240319141827.png]]<br>MyMethod3 ( ) is the only method that compiler can understand that inside the try block could throw a checked exception. <br><br>compiler doesn't care that the exception (what ever exception) is seen under the shape of the Exception.<br><br>if the compiler doesn't understand the try block  not gonna throw a checked exception even if it is seen under the shape of Exception, then compiler does not force to apply the catch or specify rule.<br><br>In the myMethod4() the exception has been specified.Also it is not necessary to throw a Exception because compiler can understand (during analyze the try block) the classNotFoundExcepiton is the type that throw in this case from the try block<br><br>Before java 7 , compiler didn't have that ability to analyze the try block.<br><br>demo - 25 |
| What is the reason to error now                                                           | ![[Pasted image 20240319144159.png]]<br><br>It is possible to specify with using the exception that throw from the try block even if the exception parameter is Exception (super exception) after java 7. But , to apply this rule the exception parameter should be ==final or effectively final==<br><br><br>![[Pasted image 20240319144635.png]]<br><br>![[Pasted image 20240319144737.png]]<br><br><br>                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
| What is effectively final                                                                 | The compiler should be understand that the variable haven't been assigned again . Even not for the null value                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  |
